### PR TITLE
[VL] Add binary, struct, decimal support for array literal

### PR DESCRIPTION
**Pull request by zhli1142015**
## What changes were proposed in this pull request?

Add binary,struct,decimal support for array lit. And do code refactor to reduce duplicate logic to match scalar types literals.

## How was this patch tested?

UT.

### CODE REVIEW

**Comment by zhli1142015** [comment]
There are another two failed native writer UTs after offload more types literals.

```
2024-01-08T00:41:31.4483691Z Caused by: org.apache.spark.SparkException: Task failed while writing rows to staging path: file:/opt/gluten/gluten-ut/spark34/target/scala-2.12/test-classes/unit-tests-working-home/spark-warehouse/tab1/_temporary/0/_temporary/attempt_202401071641318118951700050666585_0090_m_000000_123, output path: file:/opt/gluten/gluten-ut/spark34/target/scala-2.12/test-classes/unit-tests-working-home/spark-warehouse/tab1
2024-01-08T00:41:31.4486578Z 	at org.apache.spark.sql.execution.VeloxColumnarWriteFilesRDD.compute(VeloxColumnarWriteFilesExec.scala:193)
2024-01-08T00:41:31.4487630Z 	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
2024-01-08T00:41:31.4488292Z 	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
2024-01-08T00:41:31.4488970Z 	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:92)
2024-01-08T00:41:31.4489810Z 	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
2024-01-08T00:41:31.4490537Z 	at org.apache.spark.scheduler.Task.run(Task.scala:139)
2024-01-08T00:41:31.4491277Z 	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:554)
2024-01-08T00:41:31.4492153Z 	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1529)
2024-01-08T00:41:31.4492903Z 	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:557)
2024-01-08T00:41:31.4493776Z 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
2024-01-08T00:41:31.4494708Z 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
2024-01-08T00:41:31.4495400Z 	at java.lang.Thread.run(Thread.java:750)
2024-01-08T00:41:31.4496227Z Caused by: io.glutenproject.exception.GlutenException: java.lang.RuntimeException: Exception: VeloxRuntimeError
2024-01-08T00:41:31.4497035Z Error Source: RUNTIME
2024-01-08T00:41:31.4497339Z Error Code: INVALID_STATE
2024-01-08T00:41:31.4497711Z Reason: An unsupported nested encoding was found.
2024-01-08T00:41:31.4498141Z Retriable: False
2024-01-08T00:41:31.4498722Z Expression: vec.valueVector() == nullptr || vec.wrappedVector()->isFlatEncoding()
2024-01-08T00:41:31.4499332Z Function: exportFlattenedVector
2024-01-08T00:41:31.4499724Z File: ../../velox/vector/arrow/Bridge.cpp
2024-01-08T00:41:31.4500098Z Line: 830
2024-01-08T00:41:31.4500324Z Stack trace:
2024-01-08T00:41:31.4502270Z # 0  facebook::velox::VeloxException::VeloxException(char const*, unsigned long, char const*, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, bool, facebook::velox::VeloxException::Type, std::basic_string_view<char, std::char_traits<char> >)
2024-01-08T00:41:31.4504920Z # 1  void facebook::velox::detail::veloxCheckFail<facebook::velox::VeloxRuntimeError, char const*>(facebook::velox::detail::VeloxCheckFailArgs const&, char const*)
2024-01-08T00:41:31.4507141Z # 2  facebook::velox::(anonymous namespace)::exportFlattenedVector(facebook::velox::BaseVector const&, facebook::velox::(anonymous namespace)::Selection const&, ArrowOptions const&, ArrowArray&, facebook::velox::memory::MemoryPool*, facebook::velox::(anonymous namespace)::VeloxToArrowBridgeHolder&)
2024-01-08T00:41:31.4509679Z # 3  facebook::velox::(anonymous namespace)::exportToArrowImpl(facebook::velox::BaseVector const&, facebook::velox::(anonymous namespace)::Selection const&, ArrowOptions const&, ArrowArray&, facebook::velox::memory::MemoryPool*)
2024-01-08T00:41:31.4511903Z # 4  facebook::velox::(anonymous namespace)::exportToArrowImpl(facebook::velox::BaseVector const&, facebook::velox::(anonymous namespace)::Selection const&, ArrowOptions const&, ArrowArray&, facebook::velox::memory::MemoryPool*)
2024-01-08T00:41:31.4513750Z # 5  facebook::velox::exportToArrow(std::shared_ptr<facebook::velox::BaseVector> const&, ArrowArray&, facebook::velox::memory::MemoryPool*, ArrowOptions const&)
2024-01-08T00:41:31.4514968Z # 6  facebook::velox::parquet::Writer::write(std::shared_ptr<facebook::velox::BaseVector> const&)
2024-01-08T00:41:31.4516035Z # 7  facebook::velox::connector::hive::HiveDataSink::write(unsigned long, std::shared_ptr<facebook::velox::BaseVector> const&)
2024-01-08T00:41:31.4517191Z # 8  facebook::velox::connector::hive::HiveDataSink::appendData(std::shared_ptr<facebook::velox::RowVector>)
2024-01-08T00:41:31.4518203Z # 9  facebook::velox::exec::TableWriter::addInput(std::shared_ptr<facebook::velox::RowVector>)
2024-01-08T00:41:31.4519574Z # 10 facebook::velox::exec::Driver::runInternal(std::shared_ptr<facebook::velox::exec::Driver>&, std::shared_ptr<facebook::velox::exec::BlockingState>&, std::shared_ptr<facebook::velox::RowVector>&)
2024-01-08T00:41:31.4520943Z # 11 facebook::velox::exec::Driver::next(std::shared_ptr<facebook::velox::exec::BlockingState>&)
2024-01-08T00:41:31.4521722Z # 12 facebook::velox::exec::Task::next(folly::SemiFuture<folly::Unit>*)
2024-01-08T00:41:31.4522307Z # 13 gluten::WholeStageResultIterator::next()
2024-01-08T00:41:31.4522927Z # 14 Java_io_glutenproject_vectorized_ColumnarBatchOutIterator_nativeHasNext
2024-01-08T00:41:31.4523476Z # 15 0x00007f6f59250440
2024-01-08T00:41:31.4523728Z 
2024-01-08T00:41:31.4524152Z 	at io.glutenproject.vectorized.GeneralOutIterator.hasNext(GeneralOutIterator.java:39)
2024-01-08T00:41:31.4525102Z 	at scala.collection.convert.Wrappers$JIteratorWrapper.hasNext(Wrappers.scala:45)
2024-01-08T00:41:31.4525989Z 	at io.glutenproject.utils.IteratorCompleter.hasNext(Iterators.scala:69)
2024-01-08T00:41:31.4526791Z 	at io.glutenproject.utils.PayloadCloser.hasNext(Iterators.scala:35)
2024-01-08T00:41:31.4527631Z 	at io.glutenproject.utils.PipelineTimeAccumulator.hasNext(Iterators.scala:98)
2024-01-08T00:41:31.4528794Z 	at org.apache.spark.sql.execution.VeloxColumnarWriteFilesRDD.$anonfun$compute$2(VeloxColumnarWriteFilesExec.scala:174)
2024-01-08T00:41:31.4529869Z 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
2024-01-08T00:41:31.4530750Z 	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1563)
2024-01-08T00:41:31.4531906Z 	at org.apache.spark.sql.execution.VeloxColumnarWriteFilesRDD.compute(VeloxColumnarWriteFilesExec.scala:169)
2024-01-08T00:41:31.4532727Z 	... 11 more

```

And another  case for ` Field name must not be empty.`:  `SPARK-23173 Writing a file with data converted from JSON with and incorrect user schema`.

@JkSelf  FYI.

**Comment by zhli1142015** [comment]
cc @rui-mo and @PHILO-HE also, thanks.

File: cpp/velox/substrait/SubstraitToVeloxExpr.cc [hunk 1]
```diff
@@ -163,6 +180,44 @@ VectorPtr constructFlatVector(
...
+    case ::substrait::Expression_Literal::LiteralTypeCase::kI8:
+      return TINYINT();
+    case ::substrait::Expression_Literal::LiteralTypeCase::kI16:
+      return SMALLINT();
+    case ::substrait::Expression_Literal::LiteralTypeCase::kI32:
+      return INTEGER();
+    case ::substrait::Expression_Literal::LiteralTypeCase::kFp32:
+      return REAL();
+    case ::substrait::Expression_Literal::LiteralTypeCase::kI64:
+      return BIGINT();
```
Comment (rui-mo) [inline cpp/velox/substrait/SubstraitToVeloxExpr.cc:197]: Maybe exchange the order of BIGINT and REAL to improve readability.

File: cpp/velox/substrait/SubstraitToVeloxExpr.cc [hunk 2]
```diff
@@ -163,6 +180,44 @@ VectorPtr constructFlatVector(
...
+    case ::substrait::Expression_Literal::LiteralTypeCase::kBinary:
+      return VARBINARY();
+    case ::substrait::Expression_Literal::LiteralTypeCase::kDecimal: {
+      auto precision = literal.decimal().precision();
+      auto scale = literal.decimal().scale();
+      auto type = DECIMAL(precision, scale);
+      return type;
+    }
+    default:
+      return nullptr;
```
Comment (rui-mo) [inline cpp/velox/substrait/SubstraitToVeloxExpr.cc:217]: Does it make sense to throw exception here? For now the caller needs to throw for unsupported case anyway. After the change it doesn't need to check and handle invalid return value.

**Review by rui-mo — COMMENTED** [review]
Thanks for the support and refactor.

File: backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala [hunk 3]
```diff
@@ -62,4 +71,70 @@ class VeloxLiteralSuite extends VeloxWholeStageTransformerSuite {
...
+    validateOffloadResult("SELECT array(array(3, 4), array(5, 6))")
+    validateOffloadResult("SELECT array(map(1,2,3,4))")
+    validateOffloadResult("SELECT array(map('red', 1), map('green', 2))")
+    validateOffloadResult("SELECT array(struct(6, 'test1'), struct(5, 'test'))")
+  }
+
+  test("Map Literal") {
+    validateOffloadResult("SELECT map('b', 'a', 'e', 'e')")
+    validateOffloadResult("SELECT map(1D, 'a', 2D, 'e')")
+    validateOffloadResult("SELECT map(1.0, map(1, 2, 3, 4))")
```
Comment (JkSelf) [inline backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala:98]: Do we need to add map(map, map) ?

**Review by JkSelf — COMMENTED** [review]
Thanks for your great work. LGTM.

File: cpp/velox/substrait/SubstraitToVeloxExpr.cc (hunk 2 above)
Comment (zhli1142015) [reply cpp/velox/substrait/SubstraitToVeloxExpr.cc:217]: Now we have 3 callers for this function, which throw different error msgs for the invalid return value. As this function only covers scalar types, I think this is better to let callers to handle the invalid return value.

File: backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala (hunk 3 above)
Comment (zhli1142015) [reply backends-velox/src/test/scala/io/glutenproject/execution/VeloxLiteralSuite.scala:98]: Using `map` as key of a map is not allowed in spark. Thanks.

File: cpp/velox/substrait/SubstraitToVeloxExpr.cc (hunk 1 above)
Comment (zhli1142015) [reply cpp/velox/substrait/SubstraitToVeloxExpr.cc:195]: Updated, thanks.

**Comment by zhli1142015** [comment]
Looks no disk available in CI pipeline:
```
Disk Requirements:
   At least 3MB more space needed on the / filesystem.
```

**Review by JkSelf — APPROVED** [review]
LGTM.

**Comment by zhli1142015** [comment]
> There are another two failed native writer UTs after offload more types literals.
> 
> ```
> 2024-01-08T00:41:31.4483691Z Caused by: org.apache.spark.SparkException: Task failed while writing rows to staging path: file:/opt/gluten/gluten-ut/spark34/target/scala-2.12/test-classes/unit-tests-working-home/spark-warehouse/tab1/_temporary/0/_temporary/attempt_202401071641318118951700050666585_0090_m_000000_123, output path: file:/opt/gluten/gluten-ut/spark34/target/scala-2.12/test-classes/unit-tests-working-home/spark-warehouse/tab1
> 2024-01-08T00:41:31.4486578Z 	at org.apache.spark.sql.execution.VeloxColumnarWriteFilesRDD.compute(VeloxColumnarWriteFilesExec.scala:193)
> 2024-01-08T00:41:31.4487630Z 	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
> 2024-01-08T00:41:31.4488292Z 	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
> 2024-01-08T00:41:31.4488970Z 	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:92)
> 2024-01-08T00:41:31.4489810Z 	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
> 2024-01-08T00:41:31.4490537Z 	at org.apache.spark.scheduler.Task.run(Task.scala:139)
> 2024-01-08T00:41:31.4491277Z 	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:554)
> 2024-01-08T00:41:31.4492153Z 	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1529)
> 2024-01-08T00:41:31.4492903Z 	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:557)
> 2024-01-08T00:41:31.4493776Z 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 2024-01-08T00:41:31.4494708Z 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 2024-01-08T00:41:31.4495400Z 	at java.lang.Thread.run(Thread.java:750)
> 2024-01-08T00:41:31.4496227Z Caused by: io.glutenproject.exception.GlutenException: java.lang.RuntimeException: Exception: VeloxRuntimeError
> 2024-01-08T00:41:31.4497035Z Error Source: RUNTIME
> 2024-01-08T00:41:31.4497339Z Error Code: INVALID_STATE
> 2024-01-08T00:41:31.4497711Z Reason: An unsupported nested encoding was found.
> 2024-01-08T00:41:31.4498141Z Retriable: False
> 2024-01-08T00:41:31.4498722Z Expression: vec.valueVector() == nullptr || vec.wrappedVector()->isFlatEncoding()
> 2024-01-08T00:41:31.4499332Z Function: exportFlattenedVector
> 2024-01-08T00:41:31.4499724Z File: ../../velox/vector/arrow/Bridge.cpp
> 2024-01-08T00:41:31.4500098Z Line: 830
> 2024-01-08T00:41:31.4500324Z Stack trace:
> 2024-01-08T00:41:31.4502270Z # 0  facebook::velox::VeloxException::VeloxException(char const*, unsigned long, char const*, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, bool, facebook::velox::VeloxException::Type, std::basic_string_view<char, std::char_traits<char> >)
> 2024-01-08T00:41:31.4504920Z # 1  void facebook::velox::detail::veloxCheckFail<facebook::velox::VeloxRuntimeError, char const*>(facebook::velox::detail::VeloxCheckFailArgs const&, char const*)
> 2024-01-08T00:41:31.4507141Z # 2  facebook::velox::(anonymous namespace)::exportFlattenedVector(facebook::velox::BaseVector const&, facebook::velox::(anonymous namespace)::Selection const&, ArrowOptions const&, ArrowArray&, facebook::velox::memory::MemoryPool*, facebook::velox::(anonymous namespace)::VeloxToArrowBridgeHolder&)
> 2024-01-08T00:41:31.4509679Z # 3  facebook::velox::(anonymous namespace)::exportToArrowImpl(facebook::velox::BaseVector const&, facebook::velox::(anonymous namespace)::Selection const&, ArrowOptions const&, ArrowArray&, facebook::velox::memory::MemoryPool*)
> 2024-01-08T00:41:31.4511903Z # 4  facebook::velox::(anonymous namespace)::exportToArrowImpl(facebook::velox::BaseVector const&, facebook::velox::(anonymous namespace)::Selection const&, ArrowOptions const&, ArrowArray&, facebook::velox::memory::MemoryPool*)
> 2024-01-08T00:41:31.4513750Z # 5  facebook::velox::exportToArrow(std::shared_ptr<facebook::velox::BaseVector> const&, ArrowArray&, facebook::velox::memory::MemoryPool*, ArrowOptions const&)
> 2024-01-08T00:41:31.4514968Z # 6  facebook::velox::parquet::Writer::write(std::shared_ptr<facebook::velox::BaseVector> const&)
> 2024-01-08T00:41:31.4516035Z # 7  facebook::velox::connector::hive::HiveDataSink::write(unsigned long, std::shared_ptr<facebook::velox::BaseVector> const&)
> 2024-01-08T00:41:31.4517191Z # 8  facebook::velox::connector::hive::HiveDataSink::appendData(std::shared_ptr<facebook::velox::RowVector>)
> 2024-01-08T00:41:31.4518203Z # 9  facebook::velox::exec::TableWriter::addInput(std::shared_ptr<facebook::velox::RowVector>)
> 2024-01-08T00:41:31.4519574Z # 10 facebook::velox::exec::Driver::runInternal(std::shared_ptr<facebook::velox::exec::Driver>&, std::shared_ptr<facebook::velox::exec::BlockingState>&, std::shared_ptr<facebook::velox::RowVector>&)
> 2024-01-08T00:41:31.4520943Z # 11 facebook::velox::exec::Driver::next(std::shared_ptr<facebook::velox::exec::BlockingState>&)
> 2024-01-08T00:41:31.4521722Z # 12 facebook::velox::exec::Task::next(folly::SemiFuture<folly::Unit>*)
> 2024-01-08T00:41:31.4522307Z # 13 gluten::WholeStageResultIterator::next()
> 2024-01-08T00:41:31.4522927Z # 14 Java_io_glutenproject_vectorized_ColumnarBatchOutIterator_nativeHasNext
> 2024-01-08T00:41:31.4523476Z # 15 0x00007f6f59250440
> 2024-01-08T00:41:31.4523728Z 
> 2024-01-08T00:41:31.4524152Z 	at io.glutenproject.vectorized.GeneralOutIterator.hasNext(GeneralOutIterator.java:39)
> 2024-01-08T00:41:31.4525102Z 	at scala.collection.convert.Wrappers$JIteratorWrapper.hasNext(Wrappers.scala:45)
> 2024-01-08T00:41:31.4525989Z 	at io.glutenproject.utils.IteratorCompleter.hasNext(Iterators.scala:69)
> 2024-01-08T00:41:31.4526791Z 	at io.glutenproject.utils.PayloadCloser.hasNext(Iterators.scala:35)
> 2024-01-08T00:41:31.4527631Z 	at io.glutenproject.utils.PipelineTimeAccumulator.hasNext(Iterators.scala:98)
> 2024-01-08T00:41:31.4528794Z 	at org.apache.spark.sql.execution.VeloxColumnarWriteFilesRDD.$anonfun$compute$2(VeloxColumnarWriteFilesExec.scala:174)
> 2024-01-08T00:41:31.4529869Z 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
> 2024-01-08T00:41:31.4530750Z 	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1563)
> 2024-01-08T00:41:31.4531906Z 	at org.apache.spark.sql.execution.VeloxColumnarWriteFilesRDD.compute(VeloxColumnarWriteFilesExec.scala:169)
> 2024-01-08T00:41:31.4532727Z 	... 11 more
> ```
> 
> And another case for ` Field name must not be empty.`: `SPARK-23173 Writing a file with data converted from JSON with and incorrect user schema`.
> 
> @JkSelf FYI.

Thanks @JkSelf , should these two cases be updated to #3559 or #4110?

**Comment by JkSelf** [comment]
> > There are another two failed native writer UTs after offload more types literals.
> > ```
> > 2024-01-08T00:41:31.4483691Z Caused by: org.apache.spark.SparkException: Task failed while writing rows to staging path: file:/opt/gluten/gluten-ut/spark34/target/scala-2.12/test-classes/unit-tests-working-home/spark-warehouse/tab1/_temporary/0/_temporary/attempt_202401071641318118951700050666585_0090_m_000000_123, output path: file:/opt/gluten/gluten-ut/spark34/target/scala-2.12/test-classes/unit-tests-working-home/spark-warehouse/tab1
> > 2024-01-08T00:41:31.4486578Z 	at org.apache.spark.sql.execution.VeloxColumnarWriteFilesRDD.compute(VeloxColumnarWriteFilesExec.scala:193)
> > 2024-01-08T00:41:31.4487630Z 	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:364)
> > 2024-01-08T00:41:31.4488292Z 	at org.apache.spark.rdd.RDD.iterator(RDD.scala:328)
> > 2024-01-08T00:41:31.4488970Z 	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:92)
> > 2024-01-08T00:41:31.4489810Z 	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
> > 2024-01-08T00:41:31.4490537Z 	at org.apache.spark.scheduler.Task.run(Task.scala:139)
> > 2024-01-08T00:41:31.4491277Z 	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:554)
> > 2024-01-08T00:41:31.4492153Z 	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1529)
> > 2024-01-08T00:41:31.4492903Z 	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:557)
> > 2024-01-08T00:41:31.4493776Z 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> > 2024-01-08T00:41:31.4494708Z 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> > 2024-01-08T00:41:31.4495400Z 	at java.lang.Thread.run(Thread.java:750)
> > 2024-01-08T00:41:31.4496227Z Caused by: io.glutenproject.exception.GlutenException: java.lang.RuntimeException: Exception: VeloxRuntimeError
> > 2024-01-08T00:41:31.4497035Z Error Source: RUNTIME
> > 2024-01-08T00:41:31.4497339Z Error Code: INVALID_STATE
> > 2024-01-08T00:41:31.4497711Z Reason: An unsupported nested encoding was found.
> > 2024-01-08T00:41:31.4498141Z Retriable: False
> > 2024-01-08T00:41:31.4498722Z Expression: vec.valueVector() == nullptr || vec.wrappedVector()->isFlatEncoding()
> > 2024-01-08T00:41:31.4499332Z Function: exportFlattenedVector
> > 2024-01-08T00:41:31.4499724Z File: ../../velox/vector/arrow/Bridge.cpp
> > 2024-01-08T00:41:31.4500098Z Line: 830
> > 2024-01-08T00:41:31.4500324Z Stack trace:
> > 2024-01-08T00:41:31.4502270Z # 0  facebook::velox::VeloxException::VeloxException(char const*, unsigned long, char const*, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, bool, facebook::velox::VeloxException::Type, std::basic_string_view<char, std::char_traits<char> >)
> > 2024-01-08T00:41:31.4504920Z # 1  void facebook::velox::detail::veloxCheckFail<facebook::velox::VeloxRuntimeError, char const*>(facebook::velox::detail::VeloxCheckFailArgs const&, char const*)
> > 2024-01-08T00:41:31.4507141Z # 2  facebook::velox::(anonymous namespace)::exportFlattenedVector(facebook::velox::BaseVector const&, facebook::velox::(anonymous namespace)::Selection const&, ArrowOptions const&, ArrowArray&, facebook::velox::memory::MemoryPool*, facebook::velox::(anonymous namespace)::VeloxToArrowBridgeHolder&)
> > 2024-01-08T00:41:31.4509679Z # 3  facebook::velox::(anonymous namespace)::exportToArrowImpl(facebook::velox::BaseVector const&, facebook::velox::(anonymous namespace)::Selection const&, ArrowOptions const&, ArrowArray&, facebook::velox::memory::MemoryPool*)
> > 2024-01-08T00:41:31.4511903Z # 4  facebook::velox::(anonymous namespace)::exportToArrowImpl(facebook::velox::BaseVector const&, facebook::velox::(anonymous namespace)::Selection const&, ArrowOptions const&, ArrowArray&, facebook::velox::memory::MemoryPool*)
> > 2024-01-08T00:41:31.4513750Z # 5  facebook::velox::exportToArrow(std::shared_ptr<facebook::velox::BaseVector> const&, ArrowArray&, facebook::velox::memory::MemoryPool*, ArrowOptions const&)
> > 2024-01-08T00:41:31.4514968Z # 6  facebook::velox::parquet::Writer::write(std::shared_ptr<facebook::velox::BaseVector> const&)
> > 2024-01-08T00:41:31.4516035Z # 7  facebook::velox::connector::hive::HiveDataSink::write(unsigned long, std::shared_ptr<facebook::velox::BaseVector> const&)
> > 2024-01-08T00:41:31.4517191Z # 8  facebook::velox::connector::hive::HiveDataSink::appendData(std::shared_ptr<facebook::velox::RowVector>)
> > 2024-01-08T00:41:31.4518203Z # 9  facebook::velox::exec::TableWriter::addInput(std::shared_ptr<facebook::velox::RowVector>)
> > 2024-01-08T00:41:31.4519574Z # 10 facebook::velox::exec::Driver::runInternal(std::shared_ptr<facebook::velox::exec::Driver>&, std::shared_ptr<facebook::velox::exec::BlockingState>&, std::shared_ptr<facebook::velox::RowVector>&)
> > 2024-01-08T00:41:31.4520943Z # 11 facebook::velox::exec::Driver::next(std::shared_ptr<facebook::velox::exec::BlockingState>&)
> > 2024-01-08T00:41:31.4521722Z # 12 facebook::velox::exec::Task::next(folly::SemiFuture<folly::Unit>*)
> > 2024-01-08T00:41:31.4522307Z # 13 gluten::WholeStageResultIterator::next()
> > 2024-01-08T00:41:31.4522927Z # 14 Java_io_glutenproject_vectorized_ColumnarBatchOutIterator_nativeHasNext
> > 2024-01-08T00:41:31.4523476Z # 15 0x00007f6f59250440
> > 2024-01-08T00:41:31.4523728Z 
> > 2024-01-08T00:41:31.4524152Z 	at io.glutenproject.vectorized.GeneralOutIterator.hasNext(GeneralOutIterator.java:39)
> > 2024-01-08T00:41:31.4525102Z 	at scala.collection.convert.Wrappers$JIteratorWrapper.hasNext(Wrappers.scala:45)
> > 2024-01-08T00:41:31.4525989Z 	at io.glutenproject.utils.IteratorCompleter.hasNext(Iterators.scala:69)
> > 2024-01-08T00:41:31.4526791Z 	at io.glutenproject.utils.PayloadCloser.hasNext(Iterators.scala:35)
> > 2024-01-08T00:41:31.4527631Z 	at io.glutenproject.utils.PipelineTimeAccumulator.hasNext(Iterators.scala:98)
> > 2024-01-08T00:41:31.4528794Z 	at org.apache.spark.sql.execution.VeloxColumnarWriteFilesRDD.$anonfun$compute$2(VeloxColumnarWriteFilesExec.scala:174)
> > 2024-01-08T00:41:31.4529869Z 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
> > 2024-01-08T00:41:31.4530750Z 	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1563)
> > 2024-01-08T00:41:31.4531906Z 	at org.apache.spark.sql.execution.VeloxColumnarWriteFilesRDD.compute(VeloxColumnarWriteFilesExec.scala:169)
> > 2024-01-08T00:41:31.4532727Z 	... 11 more
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And another case for ` Field name must not be empty.`: `SPARK-23173 Writing a file with data converted from JSON with and incorrect user schema`.
> > @JkSelf FYI.
> 
> Thanks @JkSelf , should these two cases be updated to #3559 or #4110?

Please help to update in https://github.com/oap-project/gluten/issues/4110. Thanks.

**Pull request merged**